Term based concept sorting support

cmark opened the pull request:
* This PR adds new term based sorting functionality to the SNOMED CT Concept Search API. To accomplish this kind of sorting properly the index format needed to be changed. 
From now on a concept document will contain description fragments (id, storageKey, term, typeId, languageRefSetIds) to allow scripts to access these data for sorting.
* This PR also changes how `pt()` and `fsn()` expansion works. Expanding `pt()` or `fsn()` now won't try to look for the description in the index but use the data indexed in the description fragments array on the concept. This will improve `pt()` and `fsn()` expansion performance significantly.
* This PR also changes `referringMembers` and `referringMappingMembers` index fields. These used to track mapping and non-mapping active membership of all SNOMED CT core components to easily search for whether a component is an active member of a reference set or not. Unfortunately, these fields can't be used to search inactive membership efficiently. To support this, SNOMED CT component documents now has `memberOf` and `activeMemberOf` fields instead of `referringMembers` and `referringMappingMembers`. `memberOf` field tracks all memberships including active and inactive while `activeMemberOf` track only active memberships. The new fields track all reference set memberships in these fields, the distinction of mapping vs. non-mapping has been removed.
* This PR also adds a new CDORevisionCache implementation to reduce memory overhead when the server is using a revision cache. The new cache can be configured to evict cached revision based on time and size (will be added soon). This PR changes the default cache to this implementation `EvictingRevisionCache`.

cmark committed:
SO-2899: add basic sorting of concepts based on description terms...

...and locales preference list

Add support for script based sorting in index API.

cmark committed:
SO-2899: remove unused Script.fields() argument

cmark committed:
SO-2899: replace hardcoded SYNONYM ID with a set of synonym IDs

cmark committed:
SO-2899: fall back to first FSN in the description fragment list...

...if there is no preferred synonym or FSN found earlier based on the
selected locales.

cmark committed:
SO-2899: remove unnecessary languageCode property from desc. fragment

cmark committed:
SO-2899: index description fragments during concept change processing

Supported use cases:
* Create concept with descriptions
* Add new FSN/Synonym to existing concept
* Remove existing FSN/Synonym from existing concept
* Description update support (term, acceptability, etc.) is WIP

cmark committed:
SO-2899: detect and apply description term and status flag changes

cmark committed:
SO-2899: use storageKey to sort description by creation time...

...instead of effective time

Change sorting to use SnomedDescriptionFragment type instead of
Description CDOObjects.

cmark committed:
SO-2899: reduce size of description fragments

Remove unnecessary acceptabilityId field.
Index only a single description fragment for a single description by
collecting all preferred language refset membership into a List of
language refset IDs.

cmark committed:
SO-2899: update concept document only if there were relevant...

...description fragment changes (term, type, languagerefsetmember,
status).

cmark committed:
SO-2899: keep the previous fragments in case of non-description changes

cmark committed:
SO-2899: support script sort in SearchResourceRequest and Builders

cmark committed:
SO-2899: detect lang. member changes in ConceptChangeProcessor

cmark committed:
SO-2899: extract dirty cdo object filtering based on allowed feature set

cmark committed:
SO-2899: refactor SNOMED CT core component referring member fields

Track both mapping and non-mapping reference set IDs in the same arrays,
and also track inactive memberships (this is to help sorting work
properly when clients request inactive members as well).
Add memberOf field to track all referring members of any refset type,
both active and inactive memberships.
Add activeMemberOf field to track active referring members of any refset
type.

cmark committed:
SO-2899: add isMemberOf filter to SNOMED CT component APIs

cmark committed:
SO-2899: skip acceptable only description from being indexed

cmark committed:
SO-2899: add preferredDescription field to concept representation

Also reduce cost of expanding PT and FSN fields by using the description
fragment data available on the concept documents to calculate the
values.

cmark committed:
SO-2899: remove unnecessary fetching of all visited concepts...

...at the end of the import

Send back only the visited concept IDs.

cmark committed:
SO-2899: do not trigger concept doc update in case of dirty...

...text definitions

cmark committed:
[core] copy commit revisions before adjusting them

This fixes various inconsistency issues when using an alternative
revision cache (see next commit). Also probably this was the cause of a
few nondeterministic test failures in Jenkins.

cmark committed:
[core] use a custom time (and soon size) eviction based revision cache..

...implementation instead of a the default softreference based memory
sensitive cache.

apeteri commented on commons/com.b2international.index.api/src/com/b2international/index/Script.java:
```diff
@@ -30,20 +30,13 @@
-	 * Optionally specify the fields to be loaded for the script
-	 * @return
-	 */
-	String[] fields() default {};
```
Is this capability (to only use certain fields in a script) no longer useful to have?

cmark replied:
It was used by the Lucene index API implementation to load the listed fields for the script's execution. 
In 6.0, we've removed that module, thus this is no longer in use. 
Elasticsearch loads fields/docvalues/source on demand depending on the actual script code and it does not require fields to be explicitly named before the execution, so this fields() argument is unnecessary.

apeteri commented on commons/com.b2international.index.api/src/com/b2international/index/query/SortBy.java:
```diff
@@ -127,27 +173,25 @@ public String toString() {
+			return this;
+		}
+		
+		public Builder byScript(String script, Map<String, Object> arguments, Order order) {
```
The two builder methods should have consistent naming. Either this needs the `sort-` prefix or the other needs to be cut.

cmark replied:
Yes, thanks!

apeteri commented on commons/com.b2international.index.api/src/com/b2international/index/query/SortBy.java:
```diff
@@ -127,27 +173,25 @@ public String toString() {
-		public Builder add(String field, Order order) {
-			sortOrderMap.put(field, order);
+		public Builder sortByField(String field, Order order) {
+			sorts.add(new SortByField(field, order));
```
The factory method for the field-based `SortBy` should be used here.

apeteri commented on commons/com.b2international.index.es/src/com/b2international/index/EsDocumentSearcher.java:
```diff
@@ -254,25 +256,47 @@ public void cancelScroll(String scrollId) {
+				String script = sortByScript.getName();
+				SortOrder sortOrder = order == SortBy.Order.ASC ? SortOrder.ASC : SortOrder.DESC;
+				
+				// if this is a named script the get it from the current mapping
```
⚠️ the &rarr; then

apeteri commented on core/com.b2international.snowowl.datastore.server/src/com/b2international/snowowl/datastore/server/EvictingRevisionCache.java:
```diff
@@ -0,0 +1,368 @@
+			}
+
+			if (list.addRevision((InternalCDORevision) revision)) {
+				typeRefIncrease(id, revision.getEClass());
```
There seems to be an imbalance in the typeRef counter, here it is registering all revisions that refer to a CDOID-EClass pair...

apeteri commented on core/com.b2international.snowowl.datastore.server/src/com/b2international/snowowl/datastore/server/EvictingRevisionCache.java:
```diff
@@ -0,0 +1,368 @@
+				list.removeRevision(branchVersion.getVersion());
+				if (list.isEmpty()) {
+					revisionLists.invalidate(key);
+					typeRefDecrease(id);
```
...but here the counter is only decremented on the removal of the last revision, once.

cmark replied:
Fixed it!

apeteri commented on core/com.b2international.snowowl.datastore.server/src/com/b2international/snowowl/datastore/server/EvictingRevisionCache.java:
```diff
@@ -0,0 +1,368 @@
+		}
+
+		public synchronized boolean addRevision(InternalCDORevision revision) {
+			if (revision.getBranch().getID() == 34 && revision.getVersion() == 2 && revision.getEClass().getName().equals("Concept")) {
```
⚠️ Diagnostic code snippet

apeteri commented on core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/datastore/CDOCommitChangeSet.java:
```diff
@@ -130,17 +139,88 @@ public long getTimestamp() {
+	public <T extends CDOObject> Iterable<T> getDirtyComponents(Class<T> type, Set<EStructuralFeature> allowedFeatures) {
+		return FluentIterable.from(getDirtyComponents(type))
+			.filter(input -> {
+				final DirtyFeatureDeltaVisitor visitor = new DirtyFeatureDeltaVisitor(allowedFeatures);
```
Not sure how much it impacts performance, but if the visitor's state could be reset to "no changes" after each check, it could be created once and moved to a local variable.

cmark replied:
We can save memory that's for sure, especially during import! 😄

apeteri commented on core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/datastore/CDOCommitChangeSet.java:
```diff
@@ -130,17 +139,88 @@ public long getTimestamp() {
+		}
+		
+		private void visitDelta(CDOFeatureDelta delta) {
+			hasAllowedChanges |= allowedFeatures.contains(delta.getFeature());
```
`CDOFeatureDeltaVisitorImpl` uses an exception to jump quickly out of list feature delta visiting if not all list deltas need to be examined. While this is generally considered an anti-pattern, it is accepted as way of unwinding the call stack if other options are not available: https://blog.jooq.org/2013/04/28/rare-uses-of-a-controlflowexception/

We might want to do the same here, and then `DirtyFeatureDeltaVisitor` would be stateless. If it runs all the way through, it's a negative answer, if it throws an exception at the first feature delta which refers to an interesting feature, it's a positive answer.

cmark replied:
Sounds great! Combining it with the extract local variable refactoring (or to even a field).

apeteri commented on snomed/com.b2international.snowowl.snomed.datastore.tests/src/com/b2international/snowowl/snomed/datastore/index/change/BaseChangeProcessorTest.java:
```diff
@@ -127,6 +131,16 @@ protected final void registerSetRevisionDelta(CDOObject object, EStructuralFeatu
 	}
 	
+	protected final void registerAddRevisionDelta(CDOObject object, EStructuralFeature feature, int index, Object value) {
+		final CDOAddFeatureDeltaImpl featureDelta = new CDOAddFeatureDeltaImpl(feature, 0, value);
```
`index` is passed in as an argument, but 0 is used in the add feature delta.

apeteri commented on snomed/com.b2international.snowowl.snomed.datastore.tests/src/com/b2international/snowowl/snomed/datastore/index/change/BaseChangeProcessorTest.java:
```diff
@@ -127,6 +131,16 @@ protected final void registerSetRevisionDelta(CDOObject object, EStructuralFeatu
+	}
+	
+	protected final void registerRemoveRevisionDelta(CDOObject object, EStructuralFeature feature, int index) {
+		final CDORemoveFeatureDeltaImpl featureDelta = new CDORemoveFeatureDeltaImpl(feature, 0);
```
`index` is passed in as an argument, but 0 is used in the remove delta.

apeteri commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/converter/SnomedConceptConverter.java:
```diff
@@ -143,6 +146,19 @@ protected SnomedConcept toResource(final SnomedConceptDocument input) {
 			result.setStatedAncestorIds(input.getStatedAncestors().toArray());
 		}
+		
+		if (expand().containsKey("preferredDescriptions") || expand().containsKey(SnomedConcept.Expand.PREFERRED_TERM) || expand().containsKey("fsn()")) {
```
`SnomedConcept.Expand.FULLY_SPECIFIED_NAME` is already defined; `preferredDescriptions` could be added there as well.

apeteri commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/converter/SnomedConceptConverter.java:
```diff
@@ -206,8 +231,17 @@ private void expandPreferredTerm(List<SnomedConcept> results, final Set<String>
+		final List<SnomedDescription> synonyms = newArrayList();
+		for (SnomedConcept result : results) {
+			for (SnomedDescription description : result.getPreferredDescriptions()) {
+				if (!Concepts.FULLY_SPECIFIED_NAME.equals(description.getTypeId())) {
```
Skip `Concepts.TEXT_DEFINITION`s here as well.

apeteri commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/index/change/ConceptChangeProcessor.java:
```diff
@@ -53,38 +45,64 @@
+			.add(SnomedRefSetPackage.Literals.SNOMED_LANGUAGE_REF_SET_MEMBER__ACCEPTABILITY_ID)
+			.build();
+	
+	private static final Ordering<SnomedDescriptionFragment> DESCRIPTION_FRAGMENT_ORDER = Ordering.natural()
```
If I'm not reading it the wrong direction, this `Ordering` seems to be sorting by storage key first. But this means that the rest of the ordering steps are wasted, because the storage key is unique, so there will be no ties.

cmark replied:
You're right! Removed unnecessary type and term ordering.

apeteri commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/index/change/ReferringMemberChangeProcessor.java:
```diff
@@ -77,10 +74,8 @@ public ReferringMemberChangeProcessor(final short referencedComponentType) {
-							addChange(memberChanges, dirtyMember, MemberChangeKind.REMOVED);
-						} else if (Boolean.FALSE == oldValue && Boolean.TRUE == newValue) {
-							addChange(memberChanges, dirtyMember, MemberChangeKind.ADDED);
+						if ((Boolean.TRUE == oldValue && Boolean.FALSE == newValue) || (Boolean.FALSE == oldValue && Boolean.TRUE == newValue)) {
```
Maybe unboxing to primitive `boolean`s would be better here, just to see if any of the values where `null` for some reason.

apeteri commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/index/entry/SnomedConceptDocument.java:
```diff
@@ -520,6 +576,10 @@ public boolean isStructural() {
 		return structural;
 	}
 	
+	public List<SnomedDescriptionFragment> getDescriptions() {
```
🤔 Is there still time to change the name to `getPreferredDescriptions` or similar, to reflect it better what this field stores?

cmark replied:
Renamed it, we will import a new dataset from scratch anyways.

apeteri commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/index/change/ConceptChangeProcessor.java:
```diff
@@ -171,6 +193,25 @@ public String apply(SnomedRefSet input) {
+								newDescriptions.put(dirtyDescription.getId(), toDescriptionFragment(dirtyDescription));
+							}
+						}
+						doc.descriptions(newDescriptions.values().stream().sorted(DESCRIPTION_FRAGMENT_ORDER).collect(Collectors.toList()));
```
I will need some help understanding this part. 😕

cmark committed:
SO-2899: rename method byScript to sortByScript

Use factory methods in SortBy.Builder.
Reduce visibility of SortBy subclass constructors.

cmark committed:
SO-2899: fix comment

cmark committed:
SO-2899: fix typeDefIncrease/Decrease balance

cmark committed:
SO-2899: remove diagnostic code snippet

cmark committed:
SO-2899: fix unused index arguments in BaseChangeProcessorTest

cmark committed:
SO-2899: fix typeDefDecrease when evicting entry from cache

cmark committed:
SO-2899: use Expand option constants in SnomedConceptConverter

Also remove unused getDescriptionHelper() method.

cmark committed:
SO-2899: remove unnecessary type and term ordering

cmark committed:
SO-2899: use primitive booleans instead of wrappers

cmark committed:
SO-2899: rename SnomedConceptDocument.descriptions...

...to preferredDescriptions.

cmark committed:
SO-2899: remove unnecessary Multimap when applying description changes..

...in ConceptChangeProcessor

cmark committed:
Merge remote-tracking branch 'origin/6.x' into improvement/SO-2899-term-sorting

cmark committed:
SO-2899: fix term sort script logic

Iterate over locales first then descriptions second to match the first
preferred description according to the locale preference list.

cmark commented:
Thanks @apeteri!
